src/lib/menu/menu-directive.ts

```diff
@@ -65,6 +70,76 @@ export class MdMenu {
     this._showClickCatcher = bool;
   }
 
+  /**
+   * Focus the first item in the menu. This method is used by the menu trigger
+   * to focus the first item when the menu is opened by the ENTER key.
+   * TODO: internal
+   */
+  _focusFirstItem() {
+    this.items.first.focus();
+  }
+
+  // TODO(kara): update this when (keydown.downArrow) testability is fixed
+  // TODO: internal
+  _handleKeydown(event: KeyboardEvent): void {
+    if (event.keyCode === DOWN_ARROW) {
+      this._focusNextItem();
+    } else if (event.keyCode === UP_ARROW) {
+      this._focusPreviousItem();
+    } else if (event.keyCode === TAB) {
+      this._emitCloseEvent();
+    }
+  }
+
+  /**
+   * This emits a close event to which the trigger is subscribed. When emitted, the
+   * trigger will close the menu.
+   */
+  private _emitCloseEvent(): void {
+    this._focusedItemIndex = 0;
+    this.close.emit(null);
+  }
+
+  // When focus would shift past the start or end of the menu, wrap back to the beginning or end
+  private _wrapIfFocusLeavesMenu(): void {
+    if (this._focusedItemIndex >= this.items.length) {
+      this._focusedItemIndex = 0;
+    } else if (this._focusedItemIndex < 0) {
+      this._focusedItemIndex = this.items.length - 1;
+    }
+  }
+
+  private _focusNextItem(): void {
+    this._updateFocusedItemIndex(1);
+    this.items.toArray()[this._focusedItemIndex].focus();
+  }
+
+  private _focusPreviousItem(): void {
+    this._updateFocusedItemIndex(-1);
+    this.items.toArray()[this._focusedItemIndex].focus();
+  }
+
+  /**
+   * This method sets focus to the correct menu item, given a list of menu items and the delta
+   * between the currently focused menu item and the new menu item to be focused. It will
+   * continue to move down the list until it finds an item that is not disabled, and it will wrap
+   * if it encounters either end of the menu.
+   *
+   * @param delta the desired change in focus index
+   * @param menuItems the menu items that should be focused
+   * @private
+     */
+  private _updateFocusedItemIndex(delta: number, menuItems: MdMenuItem[] = this.items.toArray()) {
+    this._focusedItemIndex += delta;
+    this._wrapIfFocusLeavesMenu();
```

Comment:
You don't actually need the `_wrapIfFocusLeavesMenu` function at all. You can do
```ts
this._focusedItemIndex = (this._focusedItemIndex + delta + this.items.length) % this.items.length;
```